### PR TITLE
Feature/query map annotation

### PR DESCRIPTION
**Pull request by nmiyake**

### CODE REVIEW

**Comment by nmiyake** [comment]
Preview of what QueryMap annotation support looks like in http-remoting.

Once Feign releases the next release that has my QueryMap PR that merged there, this commit will add `@QueryMap` support for JAX-RS annotated interfaces (whereas my PR on Feign only adds the annotation for the default Feign contract)

**Comment by nmiyake** [comment]
@uschi2000 @markelliot now that the Feign upgrade has merged, this works as expected. Feedback welcome. Will also look at updating #71 with a simpler approach to simplify some of the decoration stuff, but that would be somewhat independent.

**Comment by nmiyake** [comment]
@uschi2000 ping for feedback. Not urgent, but can't use `@QueryMap` on JAX-RS format interfaces in standard http-remoting without adding something like this.

Updated the PR to just be the commit to add this functionality (separated from PR for simplifying some of the decorator methods).

File: http-clients/src/main/java/com/palantir/remoting/http/GuavaOptionalAwareContract.java [hunk 1]
```diff
@@ -39,50 +35,32 @@
...
-                            md.indexToExpanderClass().put(i, GuavaEmptyOptionalExpander.class);
-                        } else if (paramAnnotations.contains(QueryParam.class)) {
-                            md.indexToExpanderClass().put(i, GuavaNullOptionalExpander.class);
-                        } else if (paramAnnotations.contains(PathParam.class)) {
-                            throw new RuntimeException(String.format(
-                                    "Cannot use Guava Optionals with PathParams. (Class: %s, Method: %s, Param: arg%d)",
-                                    targetType.getName(), method.getName(), i));
-                        }
-                    }
+    protected void processMetadata(Class<?> targetType, Method method, MethodMetadata metadata) {
```
Comment (uschi2000) [inline http-clients/src/main/java/com/palantir/remoting/http/GuavaOptionalAwareContract.java:45]: is this just an equivalent rewriting?

Comment (nmiyake) [inline http-clients/src/main/java/com/palantir/remoting/http/GuavaOptionalAwareContract.java:45]: Yes, pushes up common structure/functionality to `AbstractDelegatingContract` and then just re-writes to take advantage of that and reduce duplicated code -- no functionality change.

**Comment by uschi2000** [comment]
code-wise, this looks cool

**Comment by uschi2000** [comment]
I wonder though if QueryMap is going to be a standard use-case and if it should thus be included in the standard FeignClients factories by default?

**Comment by nmiyake** [comment]
Fair -- I don't feel strongly about including it by default. However, I would like it to be in http-remoting and have the ability to include it in generated clients so that if we publish an interface that uses it, people can still use fairly standard http-remoting practices to instantiate a client from the interface.

**Comment by uschi2000** [comment]
Agreed. @melliot, any concerns?

**Comment by nmiyake** [comment]
Only thing on my end is that if we don't include as standard, then improving the "create clients with custom settings" story does become more important. The basic pattern would be something like:

```
FeignClientFactory standardFactory = FeignClients.standard();
Builder = FeignClientFactory.builder(factory);
builder.contract(new QueryMapAwareContract(standardFactory.getContract()));
return builder.build();
```

Ideally would like to combine with the commit/work to make the over-all process easier, but just want to avoid a situation where clients have to manually call `com.palantir.remoting.http.FeignClientFactory.of(Contract, Encoder, Decoder, ErrorDecoder, Function<Optional<SSLSocketFactory>, Client>)` with parameters that just happen to match the defaults by looking at source code.

**Comment by markelliot** [comment]
It's probably fine to include in the standard setup...we should probably revisit whether we're getting mileage out of Feign here, though, increasingly we seem to be working around its setup.

**Comment by nmiyake** [comment]
Cool. Would be interested to be part of the broader discussion as well, but in the mean time sounds like this PR is good as-is then?

**Comment by nmiyake** [comment]
Thanks @uschi2000 ! Would it be possible to get a new minor release for this?

**Comment by uschi2000** [comment]
Done, 0.8.1

**Pull request merged**